compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs

```diff
@@ -2107,14 +2114,45 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
     pub fn lower_const_arg(
         &self,
         const_arg: &hir::ConstArg<'tcx>,
-        feed: FeedConstTy,
+        feed: FeedConstTy<'_, 'tcx>,
     ) -> Const<'tcx> {
         let tcx = self.tcx();
 
-        if let FeedConstTy::Param(param_def_id) = feed
+        if let FeedConstTy::Param(param_def_id, args) = feed
             && let hir::ConstArgKind::Anon(anon) = &const_arg.kind
         {
-            tcx.feed_anon_const_type(anon.def_id, tcx.type_of(param_def_id));
+            let anon_const_type = tcx.type_of(param_def_id).instantiate(tcx, args);
+
+            // We must error if the instantiated type has any inference variables as we will
+            // use this type to feed the `type_of` and query results must not contain inference
+            // variables otherwise we will ICE.
+            //
+            // We also error if the type contains any regions as effectively any region will wind
+            // up as a region variable in mir borrowck. It would also be somewhat concerning if
+            // hir typeck was using equality but mir borrowck wound up using subtyping as that could
+            // result in a non-infer in hir typeck but a region variable in borrowck.
+            //
+            // FIXME(generic_const_parameter_types): Ideally we remove these errors one day when
+            // we have the ability to intermix typeck of anon const const args with the parent
+            // bodies typeck.
+            if tcx.features().generic_const_parameter_types() && anon_const_type.has_regions() {
+                let e = tcx
+                    .dcx()
+                    .span_err(const_arg.span(), "Type of const argument contains lifetimes");
+                tcx.feed_anon_const_type(anon.def_id, ty::EarlyBinder::bind(Ty::new_error(tcx, e)));
+                return ty::Const::new_error(tcx, e);
+            }
+            if anon_const_type.has_non_region_infer() {
+                let e =
+                    tcx.dcx().span_err(const_arg.span(), "Type of const argument is uninferred");
```

Comment:
```suggestion
                    tcx.dcx().span_err(const_arg.span(), "type of const argument is uninferred");
```

ditto

Reply:
(minor) I would've expected it to say something like *cannot infer type of const argument* which sounds more natural(?)

Reply:
I reworded it a bit, I didn't want to use a normal "cannot infer type of ..." because I don't really intend to keep this error around long term. Now it comes across more like a current limitation of the impl than an actual error which I prefer